File: samples/segmentation/bisenet/bisenet.py
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)

from common.model_cv import ModelCV
import numpy as np
import cv2
import argparse
import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/bisenet-int8-kl_divergence-1_3_512_512-vacc/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/bisenet_bgr888.json ",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/face.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./bisenet_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="dataset filelist",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="dataset root",
    )
    parser.add_argument(
        "--dataset_output_folder",
        default="",
        help="dataset output folder",
    )
    args = parser.parse_args()
    return args


part_colors = [
    [0, 0, 0],
    [255, 85, 0],
    [255, 170, 0],
    [255, 0, 85],
    [255, 0, 170],
    [0, 255, 0],
    [85, 255, 0],
    [170, 255, 0],
    [0, 255, 85],
    [0, 255, 170],
    [0, 0, 255],
    [85, 0, 255],
    [170, 0, 255],
    [0, 85, 255],
    [0, 170, 255],
    [255, 255, 0],
    [255, 255, 85],
    [255, 255, 170],
    [255, 0, 255],
    [255, 85, 255],
    [255, 170, 255],
    [0, 255, 255],
    [85, 255, 255],
    [170, 255, 255],
]


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    model = ModelCV(
        args.model_prefix, args.vdsp_params, batch_size, args.device_id, args.hw_config
    )
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_filelist == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Read image failed: {args.input_image}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)

        result = model.process(vsx_image)
        class_map = np.array(result).squeeze().argmax(0)
        num_of_class = np.max(class_map)
        class_map = cv2.resize(
            class_map,
            (cv_image.shape[1], cv_image.shape[0]),
            interpolation=cv2.INTER_NEAREST,
        )
        segmentation_image = np.zeros(cv_image.shape, dtype=np.uint8)
        for pi in range(1, num_of_class + 1):
            index = np.where(class_map == pi)
            segmentation_image[index[0], index[1], :] = part_colors[pi]
        cv2.imwrite(args.output_file, segmentation_image)
        print(f"Write result to: {args.output_file}")
    else:
        filelist = []
        with open(args.dataset_filelist, "rt") as f:
            filelist = f.readlines()
        for image_file in filelist:
            image_file = os.path.join(
                args.dataset_root, image_file.strip(" ").strip("\n")
            )
            print(image_file)
            cv_image = cv2.imread(image_file)
            assert cv_image is not None, f"Read image failed: {image_file}"
            vsx_image = utils.cv_bgr888_to_vsximage(
                cv_image, image_format, args.device_id
            )

            result = model.process(vsx_image)

            base_name = os.path.basename(image_file).split(".")[0]
            npz_file = os.path.join(args.dataset_output_folder, base_name + ".npz")
            np.savez(npz_file, output_0=result[0])

File: samples/segmentation/bisenet/bisenet.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <chrono>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/segmentator.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "bisenet-int8-kl_divergence-1_3_512_512-vacc/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/bisenet_bgr888.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/face.jpg");
  args.add<std::string>("output_file", '\0', "output file", false,
                        "bisenet_result.jpg");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "dataset root", false, "");
  args.add<std::string>("dataset_output_folder", '\0', "dataset output folder",
                        false, "");
  args.parse_check(argc, argv);
  return args;
}

std::vector<cv::Vec3b> part_colors = {
    {0, 0, 0},       {255, 85, 0},    {255, 170, 0},  {255, 0, 85},
    {255, 0, 170},   {0, 255, 0},     {85, 255, 0},   {170, 255, 0},
    {0, 255, 85},    {0, 255, 170},   {0, 0, 255},    {85, 0, 255},
    {170, 0, 255},   {0, 85, 255},    {0, 170, 255},  {255, 255, 0},
    {255, 255, 85},  {255, 255, 170}, {255, 0, 255},  {255, 85, 255},
    {255, 170, 255}, {0, 255, 255},   {85, 255, 255}, {170, 255, 255},
};
int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  const int batch_size = 1;
  vsx::Segmentator segment(args.get<std::string>("model_prefix"),
                           args.get<std::string>("vdsp_params"), batch_size,
                           args.get<uint32_t>("device_id"));
  auto image_format = segment.GetFusionOpIimageFormat();

  if (!args.get<std::string>("dataset_filelist").empty()) {
    std::vector<std::string> filelist =
        vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
    CHECK(filelist.size()) << "No file in "
                           << args.get<std::string>("dataset_filelist");
    auto dataset_root = args.get<std::string>("dataset_root");
    auto dataset_output_folder = args.get<std::string>("dataset_output_folder");
    for (size_t s = 0; s < filelist.size(); s++) {
      auto filename = filelist[s];
      if (!dataset_root.empty()) filename = dataset_root + "/" + filelist[s];
      std::cout << filename << std::endl;
      auto cv_image = cv::imread(filename);
      vsx::Image vsx_image;
      CHECK(vsx::MakeVsxImage(cv_image, vsx_image, image_format) == 0);
      auto result = segment.Process(vsx_image);
      std::filesystem::path p(filename);
      std::string basename = p.stem().string();
      auto outfile = dataset_output_folder + "/" + basename + ".npz";
      std::unordered_map<std::string, vsx::Tensor> output_map;
      output_map["output_0"] = result;
      vsx::SaveTensorMap(outfile, output_map);
    }
  } else {
    auto cv_image = cv::imread(args.get<std::string>("input_file"));
    CHECK(!cv_image.empty())
        << "Failed to read image:" << args.get<std::string>("input_file")
        << std::endl;
    vsx::Image vsx_image;
    CHECK(vsx::MakeVsxImage(cv_image, vsx_image, image_format) == 0);
    auto output_tensor = segment.Process(vsx_image);
    auto result = segment.PostProcess(output_tensor);
    auto shape = result.Shape();
    int width = shape[2];
    int height = shape[1];
    cv::Mat out_mat(height, width, CV_8UC3);
    const uchar* type_data = result.Data<uchar>();
    for (int h = 0; h < height; h++) {
      for (int w = 0; w < width; w++) {
        out_mat.at<cv::Vec3b>(h, w) = part_colors[type_data[h * width + w]];
      }
    }
    cv::resize(out_mat, out_mat, cv_image.size());
    cv::imwrite(args.get<std::string>("output_file"), out_mat);
    std::cout << "Write result to: " << args.get<std::string>("output_file")
              << std::endl;
  }

  return 0;
}
